pandas/_libs/index.pyx

```diff
@@ -606,25 +606,112 @@ cdef class BaseMultiIndexCodesEngine:
-            indexer = indexer[order]
-        else:
-            indexer = self._base.get_indexer(self, lab_ints)
+    @staticmethod
+    def get_indexer_and_fill(object values, object target,
+                             object method, object limit = None) -> np.ndarray:
+        """
+        Gets an indexer, i.e. a set of indexes into `values`, for the values in
+        `target`, where the index value.
 
-        return indexer
+        If method is "backfill" then the index for a value in `target` which
+        does not exist in `values` is the index of the next match, or -1 is the
+        value is larger than the largest value in `values`.
+
+        Similarly, if the method if "pad" then the index for a value in `target`
+        which does not exist in `values` is the index of the previous match, or
+        -1 if the value is smaller then the largest value in `values`.
+
+        Parameters
+        ----------
+        values : list-like of tuples
+            must be sorted and all have the same length
+        target: list-like of tuples
+            need not be sorted, but all must have the same length, which must be
+            the same as the length of all tuples in `values`
+        method: string
+            "backfill" or "pad"
+        limit: int, optional
+            if provided, limit the number of fills to this value
+
+        Returns
+        -------
+        np.ndarray[int64_t, ndim=1] of the indexer of `target` into `values`,
+        filled with the `method` (and optionally `limit`) specified
+        """
+        if method not in ("backfill", "pad"):
+            raise ValueError(
+                f"{method} is not a valid method value; only 'backfill' and "
+                 "'pad' filling methods are supported")
+
+        cdef:
+            int64_t i, j, next_code
+            int64_t num_values, num_target_values
+            ndarray[int64_t, ndim=1] target_order
+            ndarray[object, ndim=1] target_values
+            ndarray[int64_t, ndim=1] new_codes, new_target_codes
+            ndarray[int64_t, ndim=1] sorted_indexer
+
+        target_order = np.argsort(target.values)
+        target_values = target.values[target_order]
+        num_values, num_target_values = len(values), len(target_values)
+        new_codes, new_target_codes = \
+            np.empty((num_values,)).astype('int64'), \
+            np.empty((num_target_values,)).astype('int64')
+
+        # `values` and `target_values` are both sorted, so we walk through them
+        # and memoize the (ordered) set of indices in the (implicit) merged-and
+        # sorted list of the two which belong to each of them
+        # the effect of this is to create a factorization for the (sorted)
+        # merger of the index values, where `new_codes` and `new_target_codes`
+        # are the subset of the factors which appear in `values` and `target`,
+        # respectively
+        i, j, next_code = 0, 0, 0
```

Comment:
example of how this code works:

if `values` is
```
[
    (1, 1),
    (1, 3),
    (2, 1),
    (2, 3),
]
```

and `target_values` (i.e. `target.values`, sorted) is:
```
[
   (0, 1),
   (1, 0),
   (1, 3),
   (2, 2),
   (2, 5),
]
```

then, if you merged these into a sorted list of (unique) values, you would get (including index values here, and an X indicates membership):
```
          values  target_values

0: (0, 1)               X
1: (1, 0)               X
2: (1, 1)    X
3: (1, 3)    X          X
4: (2, 1)    X
5: (2, 2)               X
6: (2, 3)    X
7: (2, 5)               X
```
and so we can think of this as a factorization of the merged and sorted list of tuples.  the (ordered) subset of this belonging to `values`, which becomes `new_codes`, is:
```
[2, 3, 4, 6]
```
and the equivalent value for `target_values`, which becomes `new_target_codes`, is:
```
[0, 1, 3, 5, 7]
```

thus, if we are getting a backfilled indexer of `target_values` into `values`, we can simply backfill `new_target_codes` into `new_codes`, ge
tting us:
```
[0, 0, 1, 3, -1]
```
similarly, for padding, we can simply call `algos.pad(new_codes, new_target_codes)`, getting us:
```
[-1, -1, 1, 2, 3]
```
which can be seen to be correct